src/google-classroom-extractor/google_classroom_extractor/api/courses.py

```diff
@@ -0,0 +1,40 @@
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+import logging
+from typing import List, Dict
+import pandas as pd
+from .api_caller import call_api
+
+
+def request_courses(resource) -> List[Dict[str, str]]:
+    # type: ignore
+    return call_api(
+        resource.courses().list,
+        {"courseStates": ["ACTIVE"]}, # type: ignore
+        "courses",
+    )
+
+
+def request_latest_courses_as_df(resource) -> pd.DataFrame:
+    logging.info("Pulling course data")
+    courses: List[Dict[str, str]] = request_courses(resource)
+    return pd.json_normalize(courses)
+
+
+def request_all_courses_as_df(resource, sync_db) -> pd.DataFrame:
```

Comment:
Should we add asserts for checking 'None' values in some of the functions?
If yes, we should unit test the none values too.

Reply:
Type annotations handle this better, but of course the parameters actually need to be typed.  I'll see if I can find the actual types for resource and sync_db, otherwise asserts are a good idea for untyped values.

Reply:
With or without type annotations, can the system end up with a `None` value on a property before it is dereferenced? If so, I like to guard against that value with an assertion, instead of having an [AttributeError](https://docs.python.org/3/library/exceptions.html#AttributeError) raised.

Reply:
@stephenfuqua @lalrg  I've been thinking about this and it's really quite interesting.  Asserting not null is a useful run-time check in statically-typed languages where null is valid for any object type (like C# before 8.0 introduced the nullable reference type option).  However, in dynamically typed languages like Python, JS, etc. it's actually a rather weak assertion.  Better assertions here would be `assert isinstance(resource, googleapiclient.discovery.Resource)` and `assert isinstance(sync_db, sqlalchemy.engine.base.Engine)`  Python types are not nullable, so these provide a stronger run-time check that it's not just "not None", but actually the type we expect.  Go with this?

Reply:
FYI, I have a duck-typing example in the Google extractor where all I expect from a "executable_resource" parameter is an object with an "execute" method.  I'm asserting now like so: `assert hasattr(executable_resource, "execute")`

Reply:
You raise a good point. Much better to do `assert type(s) is str` than `assert str is not None`.

Reply:
I have read that most python devs prefer `assert hasattr` over `assert type(s)`

Reply:
It depends on what you're trying to accomplish. Most of the time I'm just trying to see if something is a `str` type. Checking to see if it has a particular string attribute would not prove that it is a string. I don't think I've even see that pattern yet - this thread is the first time I have come across `hasattr` unless I just skimmed past it somewhere.